include/gatekeeper_gk.h

```diff
@@ -24,13 +24,113 @@
 #include "gatekeeper_ipip.h"
 #include "gatekeeper_ggu.h"
 #include "gatekeeper_mailbox.h"
+#include "gatekeeper_lpm.h"
+
+/* The maximum number of gateways that Gatekeeper may use. */
+#define GK_MAX_NUM_GW       (8)
+
+/*
+ * The LPM supports 8-bit next hop,
+ * which at most has 128 different values.
+ */
+#define GK_MAX_NUM_POLICIES (128)
 
 /*
  * A flow entry can be in one of three states:
  * request, granted, or declined.
  */
 enum gk_flow_state { GK_REQUEST, GK_GRANTED, GK_DECLINED };
 
+enum gk_policy_action {
+	/* Forward the packe to the corresponding gateway. */
+	GK_FWD_GATEWAY,
+
+	/*
+	 * The Grantor server is a neighbor.
+	 * Forward the packe to the Grantor server directly.
+	 */
+	GK_FWD_NEIGHBOR,
+
+	/* Forward the packet to the back interface. */
+	GK_FWD_BACK_NET,
+
+	/* Drop the packet. */
+	GK_DROP,
+
+	/* Invalid policy action. */
+	GK_POLICY_MAX,
+};
+
+struct gk_gateway {
+
+	uint16_t proto;
+
+	union {
+		struct in_addr  v4;
+		struct in6_addr v6;
+	} ip;
+
+	/* Cache the whole Ethernet header. */
+	struct ether_hdr eth_hdr;
+
+	/* Indicate whether the MAC address is stale or not. */
+	volatile bool stale;
```

Comment:
I'd put field `stale` before `eth_hdr`. The reasoning is the following: one always need to read `stale` first, so while reading it from memory, the cache will also bring `eth_hdr`, or at least part of it since the cache reads cache lines and not exact fields.